src/linear_algebra/symplectic_group.lean

```diff
@@ -0,0 +1,232 @@
+  A ∈ symplectic_group l R ↔ A ⬝ (J l R) ⬝ Aᵀ = J l R :=
+by simp [symplectic_group]
+
+instance coe_matrix : has_coe (symplectic_group l R) (matrix (l ⊕ l) (l ⊕ l)  R) := ⟨subtype.val⟩
+
+section coe_lemmas
+
+variables (A B : symplectic_group l R)
+
+@[simp] lemma mul_val : ↑(A * B) = (A : matrix (l ⊕ l) (l ⊕ l) R) ⬝ (B : matrix (l ⊕ l) (l ⊕ l) R)
+:= rfl
+
+@[simp] lemma one_val : ↑(1 : symplectic_group l R) = (1 : matrix (l ⊕ l) (l ⊕ l)  R) := rfl
+
+lemma mul_mem {A B : matrix (l ⊕ l) (l ⊕ l) R}
+  (hA : A ∈ symplectic_group l R) (hB : B ∈ symplectic_group l R) :
+  A ⬝ B ∈ symplectic_group l R :=
+submonoid.mul_mem _ hA hB
+
+end coe_lemmas
+
+section symplectic_J
+
+variables (l) (R)
+
+lemma J_mem : (J l R) ∈ symplectic_group l R :=
+begin
+  rw mem_iff,
+  unfold J,
+  rw [from_blocks_multiply, from_blocks_transpose, from_blocks_multiply],
+  simp,
+end
+
+/-- The canonical skew-symmetric matrix as an element in the symplectic group. -/
+def sym_J : symplectic_group l R := ⟨J l R, J_mem l R⟩
+
+variables {l} {R}
+
+@[simp] lemma coe_J : ↑(sym_J l R) = J l R := rfl
+
+end symplectic_J
+
+variables {R} {A : matrix (l ⊕ l) (l ⊕ l) R}
+
+lemma neg_mem (h : A ∈ symplectic_group l R) : -A ∈ symplectic_group l R :=
+begin
+  rw mem_iff at h ⊢,
+  simp [h],
+end
+
+variables [no_zero_divisors R] [nontrivial R]
+variables [fact (ring_char R ≠ 2)]
+
+lemma symplectic_det (hA : A ∈ symplectic_group l R) : is_unit $ det A :=
+begin
+  rw is_unit_iff_exists_inv,
+  use A.det,
+  refine (J_det_unit l R).mul_left_cancel _,
+  rw [mul_one],
+  rw mem_iff at hA,
+  apply_fun det at hA,
+  simp only [det_mul, det_transpose] at hA,
+  rw [mul_comm A.det, mul_assoc] at hA,
+  exact hA,
+end
+
+lemma transpose_mem (hA : A ∈ symplectic_group l R) :
+  Aᵀ ∈ symplectic_group l R :=
+begin
+  rw mem_iff at ⊢ hA,
+  rw transpose_transpose,
+  have huA := symplectic_det hA,
+  have huAT : is_unit (Aᵀ).det :=
+  begin
+    rw matrix.det_transpose,
+    exact huA,
+  end,
+  calc Aᵀ ⬝ J l R ⬝ A
+      = - Aᵀ ⬝ (J l R)⁻¹ ⬝ A  : by {rw J_inv, simp}
+  ... = - Aᵀ ⬝ (A ⬝ J l R ⬝ Aᵀ)⁻¹ ⬝ A : by rw hA
+  ... = - (Aᵀ ⬝ (Aᵀ⁻¹ ⬝ (J l R)⁻¹)) ⬝ A⁻¹ ⬝ A : by simp only [matrix.mul_inv_rev,
+                                                              matrix.mul_assoc, matrix.neg_mul]
+  ... = - (J l R)⁻¹ : by rw [mul_nonsing_inv_cancel_left _ _ huAT,
+                             nonsing_inv_mul_cancel_right _ _ huA]
+  ... = (J l R) : by simp [J_inv]
+end
+
+lemma transpose_mem_iff : Aᵀ ∈ symplectic_group l R ↔ A ∈ symplectic_group l R :=
+⟨λ hA, by simpa using transpose_mem hA , transpose_mem⟩
+
+lemma mem_iff' : A ∈ symplectic_group l R ↔ Aᵀ ⬝ (J l R) ⬝ A = J l R :=
+by rw [←transpose_mem_iff, mem_iff, transpose_transpose]
+
+instance : has_inv (symplectic_group l R) :=
+{ inv := λ A, ⟨- (J l R) ⬝ (A : matrix (l ⊕ l) (l ⊕ l) R)ᵀ ⬝ (J l R),
+  mul_mem (mul_mem (neg_mem $ J_mem _ _) $ transpose_mem A.2) $ J_mem _ _⟩ }
+
+@[simp] lemma coe_inv (A : symplectic_group l R) :
```

Comment:
Is it true that the inverse coincides with the matrix Inverse?

Reply:
yes, we should probably add lemmas for that as well. We wanted the inverse in `symplectic_group` to be computable and therefore we defined it by the explicit formula.

Reply:
Yes, the definition is totally reasonable (and I think matches what we do for other matrix groups); but we should have both lemmas.

Reply:
I added the lemma that says `A⁻¹ = - (J l R) ⬝ Aᵀ ⬝ (J l R)` for matrices in the symplectic group.

Reply:
I'd like to see the lemma that says `(↑(A⁻¹) : matrix _ _ _) = A⁻¹` as well; I think that lemma should be called `coe_inv'` or `coe_inv_eq_inv`.

Reply:
Added!